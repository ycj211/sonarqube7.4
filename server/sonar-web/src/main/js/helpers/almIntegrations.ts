/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { isLoggedIn } from './users';
import { CurrentUser, AlmOrganization } from '../app/types';

export function hasAdvancedALMIntegration(user: CurrentUser) {
  return (
    isLoggedIn(user) && (isBitbucket(user.externalProvider) || isGithub(user.externalProvider))
  );
}

export function isBitbucket(almKey?: string) {
  return almKey && almKey.startsWith('bitbucket');
}

export function isGithub(almKey?: string) {
  return almKey === 'github';
}

export function isVSTS(almKey?: string) {
  return almKey === 'microsoft';
}

export function isPersonal(organization?: AlmOrganization) {
  return Boolean(organization && organization.personal);
}

export function sanitizeAlmId(almKey?: string) {
  if (isBitbucket(almKey)) {
    return 'bitbucket';
  }
  return almKey;
}
