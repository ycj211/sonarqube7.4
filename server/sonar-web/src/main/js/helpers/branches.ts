/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { sortBy } from 'lodash';
import {
  BranchLike,
  Branch,
  BranchType,
  ShortLivingBranch,
  LongLivingBranch,
  PullRequest,
  MainBranch,
  BranchParameters
} from '../app/types';

export function isBranch(branchLike?: BranchLike): branchLike is Branch {
  return branchLike !== undefined && (branchLike as Branch).isMain !== undefined;
}

export function isShortLivingBranch(branchLike?: BranchLike): branchLike is ShortLivingBranch {
  return (
    isBranch(branchLike) &&
    !branchLike.isMain &&
    (branchLike as ShortLivingBranch).type === BranchType.SHORT
  );
}

export function isLongLivingBranch(branchLike?: BranchLike): branchLike is LongLivingBranch {
  return (
    isBranch(branchLike) &&
    !branchLike.isMain &&
    (branchLike as LongLivingBranch).type === BranchType.LONG
  );
}

export function isMainBranch(branchLike?: BranchLike): branchLike is MainBranch {
  return isBranch(branchLike) && branchLike.isMain;
}

export function isPullRequest(branchLike?: BranchLike): branchLike is PullRequest {
  return branchLike !== undefined && (branchLike as PullRequest).key !== undefined;
}

export function getPullRequestDisplayName(pullRequest: PullRequest) {
  return `${pullRequest.key} â€“ ${pullRequest.title}`;
}

export function getBranchLikeDisplayName(branchLike: BranchLike) {
  return isPullRequest(branchLike) ? getPullRequestDisplayName(branchLike) : branchLike.name;
}

export function getBranchLikeKey(branchLike: BranchLike) {
  return isPullRequest(branchLike) ? `pull-request-${branchLike.key}` : `branch-${branchLike.name}`;
}

export function getBranchQualityGateColor(status: string) {
  let indicatorColor = 'gray';
  if (status === 'ERROR') {
    indicatorColor = 'red';
  } else if (status === 'WARN') {
    indicatorColor = 'orange';
  } else if (status === 'OK') {
    indicatorColor = 'green';
  }
  return indicatorColor;
}

export function isSameBranchLike(a: BranchLike | undefined, b: BranchLike | undefined) {
  // main branches are always equal
  if (isMainBranch(a) && isMainBranch(b)) {
    return true;
  }

  // short- and long-living branches are compared by type and name
  if (
    (isLongLivingBranch(a) && isLongLivingBranch(b)) ||
    (isShortLivingBranch(a) && isShortLivingBranch(b))
  ) {
    return a.type === b.type && a.name === b.name;
  }

  // pull requests are compared by id
  if (isPullRequest(a) && isPullRequest(b)) {
    return a.key === b.key;
  }

  // finally if both parameters are `undefined`, consider them equal
  return a === b;
}

export function sortBranchesAsTree(branchLikes: BranchLike[]) {
  const result: BranchLike[] = [];

  const mainBranch = branchLikes.find(isMainBranch);
  const longLivingBranches = branchLikes.filter(isLongLivingBranch);
  const shortLivingBranches = branchLikes.filter(isShortLivingBranch);
  const pullRequests = branchLikes.filter(isPullRequest);

  // main branch is always first
  if (mainBranch) {
    result.push(
      mainBranch,
      ...getPullRequests(mainBranch.name),
      ...getNestedShortLivingBranches(mainBranch.name)
    );
  }

  // then all long-living branches
  sortBy(longLivingBranches, 'name').forEach(longLivingBranch => {
    result.push(
      longLivingBranch,
      ...getPullRequests(longLivingBranch.name),
      ...getNestedShortLivingBranches(longLivingBranch.name)
    );
  });

  // finally all orhpan pull requests and branches
  result.push(
    ...sortBy(pullRequests.filter(pr => pr.isOrphan), pullRequest => pullRequest.key),
    ...sortBy(shortLivingBranches.filter(branch => branch.isOrphan), branch => branch.name)
  );

  return result;

  /** Get all short-living branches (possibly nested) which should be merged to a given branch */
  function getNestedShortLivingBranches(mergeBranch: string) {
    const found: ShortLivingBranch[] = shortLivingBranches.filter(
      branch => branch.mergeBranch === mergeBranch
    );

    let i = 0;
    while (i < found.length) {
      const current = found[i];
      found.push(...shortLivingBranches.filter(branch => branch.mergeBranch === current.name));
      i++;
    }

    return sortBy(found, branch => branch.name);
  }

  function getPullRequests(base: string) {
    return sortBy(pullRequests.filter(pr => pr.base === base), pullRequest => pullRequest.key);
  }
}

export function getBranchLikeQuery(branchLike?: BranchLike): BranchParameters {
  if (isShortLivingBranch(branchLike) || isLongLivingBranch(branchLike)) {
    return { branch: branchLike.name };
  } else if (isPullRequest(branchLike)) {
    return { pullRequest: branchLike.key };
  } else {
    return {};
  }
}

// Create branch object from branch name or pull request key
export function fillBranchLike(
  branch?: string,
  pullRequest?: string
): ShortLivingBranch | PullRequest | undefined {
  if (branch) {
    return {
      isMain: false,
      mergeBranch: '',
      name: branch,
      type: BranchType.SHORT
    } as ShortLivingBranch;
  } else if (pullRequest) {
    return { base: '', branch: '', key: pullRequest, title: '' } as PullRequest;
  }
  return undefined;
}
