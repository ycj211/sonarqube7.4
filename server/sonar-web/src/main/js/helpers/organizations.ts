/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { isLoggedIn } from './users';
import { Organization, OrganizationSubscription, CurrentUser } from '../app/types';

export function isPaidOrganization(organization: Organization | undefined): boolean {
  return Boolean(organization && organization.subscription === OrganizationSubscription.Paid);
}

export function hasPrivateAccess(
  currentUser: CurrentUser,
  organization: Organization | undefined,
  userOrganizations: Organization[]
): boolean {
  return (
    !isPaidOrganization(organization) ||
    isCurrentUserMemberOf(currentUser, organization, userOrganizations)
  );
}

export function isCurrentUserMemberOf(
  currentUser: CurrentUser,
  organization: Organization | undefined,
  userOrganizations: Organization[]
): boolean {
  return Boolean(
    organization &&
      isLoggedIn(currentUser) &&
      ((organization.actions && organization.actions.admin) ||
        userOrganizations.some(org => org.key === organization.key))
  );
}
