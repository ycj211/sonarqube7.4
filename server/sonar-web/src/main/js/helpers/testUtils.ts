/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow, ShallowRendererProps, ShallowWrapper, ReactWrapper } from 'enzyme';
import { InjectedRouter } from 'react-router';
import { IntlProvider } from 'react-intl';

export const mockEvent = {
  target: { blur() {} },
  currentTarget: { blur() {} },
  preventDefault() {},
  stopPropagation() {}
};

export function click(element: ShallowWrapper | ReactWrapper, event = {}): void {
  // `type()` returns a component constructor for a composite element and string for DOM nodes
  if (typeof element.type() === 'function') {
    element.prop<Function>('onClick')();
    // TODO find out if `root` is a public api
    // https://github.com/airbnb/enzyme/blob/master/packages/enzyme/src/ReactWrapper.js#L109
    (element as any).root().update();
  } else {
    element.simulate('click', { ...mockEvent, ...event });
  }
}

export function clickOutside(event = {}): void {
  const dispatchedEvent = new MouseEvent('click', event);
  window.dispatchEvent(dispatchedEvent);
}

export function submit(element: ShallowWrapper | ReactWrapper): void {
  element.simulate('submit', {
    preventDefault() {}
  });
}

export function change(element: ShallowWrapper | ReactWrapper, value: string, event = {}): void {
  // `type()` returns a component constructor for a composite element and string for DOM nodes
  if (typeof element.type() === 'function') {
    element.prop<Function>('onChange')(value);
    // TODO find out if `root` is a public api
    // https://github.com/airbnb/enzyme/blob/master/packages/enzyme/src/ReactWrapper.js#L109
    (element as any).root().update();
  } else {
    element.simulate('change', {
      target: { value },
      currentTarget: { value },
      ...event
    });
  }
}

export function keydown(keyCode: number): void {
  const event = new KeyboardEvent('keydown', { keyCode } as KeyboardEventInit);
  document.dispatchEvent(event);
}

export function elementKeydown(element: ShallowWrapper, keyCode: number): void {
  const event = {
    currentTarget: { element },
    keyCode,
    preventDefault() {}
  };

  if (typeof element.type() === 'string') {
    // `type()` is string for native dom elements
    element.simulate('keydown', event);
  } else {
    element.prop<Function>('onKeyDown')(event);
  }
}

export function resizeWindowTo(width?: number, height?: number) {
  // `document.body.clientWidth/clientHeight` are getters by default, so we need to redefine them
  // pass `configurable: true` to allow to redefine the properties multiple times
  if (width) {
    Object.defineProperty(document.body, 'clientWidth', { configurable: true, value: width });
  }
  if (height) {
    Object.defineProperty(document.body, 'clientHeight', { configurable: true, value: height });
  }

  const resizeEvent = new Event('resize');
  window.dispatchEvent(resizeEvent);
}

export function doAsync(fn?: Function): Promise<void> {
  return new Promise(resolve => {
    setImmediate(() => {
      if (fn) {
        fn();
      }
      resolve();
    });
  });
}

// Create the IntlProvider to retrieve context for wrapping around.
const intlProvider = new IntlProvider({ locale: 'en' }, {});
const { intl } = intlProvider.getChildContext();
export function shallowWithIntl<C extends React.Component>(
  node: React.ReactElement<any>,
  options: ShallowRendererProps = {}
) {
  return shallow<C>(node, { ...options, context: { intl, ...options.context } });
}

export async function waitAndUpdate(wrapper: ShallowWrapper<any, any> | ReactWrapper<any, any>) {
  await new Promise(setImmediate);
  wrapper.update();
}

export function mockRouter(overrides: { push?: Function; replace?: Function } = {}) {
  return {
    createHref: jest.fn(),
    createPath: jest.fn(),
    go: jest.fn(),
    goBack: jest.fn(),
    goForward: jest.fn(),
    isActive: jest.fn(),
    push: jest.fn(),
    replace: jest.fn(),
    setRouteLeaveHook: jest.fn(),
    ...overrides
  } as InjectedRouter;
}
