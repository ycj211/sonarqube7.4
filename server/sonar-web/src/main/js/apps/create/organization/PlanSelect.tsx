/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { FormattedMessage } from 'react-intl';
import { Link } from 'react-router';
import Radio from '../../../components/controls/Radio';
import { translate } from '../../../helpers/l10n';

export enum Plan {
  Free = 'free',
  Paid = 'paid'
}

interface Props {
  onChange: (plan: Plan) => void;
  plan: Plan;
  startingPrice: string;
}

export default class PlanSelect extends React.PureComponent<Props> {
  handleFreePlanClick = () => {
    this.props.onChange(Plan.Free);
  };

  handlePaidPlanClick = () => {
    this.props.onChange(Plan.Paid);
  };

  render() {
    const { plan } = this.props;
    return (
      <div
        aria-label={translate('onboarding.create_organization.choose_plan')}
        className="huge-spacer-bottom"
        role="radiogroup">
        <div>
          <Radio checked={plan === Plan.Free} onCheck={this.handleFreePlanClick}>
            <span>{translate('billing.free_plan.title')}</span>
          </Radio>
          <p className="note markdown little-spacer-top">
            {translate('billing.free_plan.description')}
          </p>
        </div>
        <div className="big-spacer-top">
          <Radio checked={plan === Plan.Paid} onCheck={this.handlePaidPlanClick}>
            <span>{translate('billing.paid_plan.title')}</span>
          </Radio>
          <p className="note markdown little-spacer-top">
            <FormattedMessage
              defaultMessage={translate('billing.paid_plan.description')}
              id="billing.paid_plan.description"
              values={{
                price: this.props.startingPrice,
                more: (
                  <>
                    {' '}
                    <Link target="_blank" to="/documentation/sonarcloud-pricing/">
                      {translate('learn_more')}
                    </Link>
                    <br />
                  </>
                )
              }}
            />
          </p>
        </div>
      </div>
    );
  }
}
