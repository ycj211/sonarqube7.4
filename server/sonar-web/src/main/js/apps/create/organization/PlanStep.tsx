/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import BillingFormShim from './BillingFormShim';
import PlanSelect, { Plan } from './PlanSelect';
import Step from '../../tutorials/components/Step';
import { withCurrentUser } from '../../../components/hoc/withCurrentUser';
import { translate } from '../../../helpers/l10n';
import { getExtensionStart } from '../../../app/components/extensions/utils';
import { SubscriptionPlan } from '../../../app/types';
import { SubmitButton } from '../../../components/ui/buttons';
import DeferredSpinner from '../../../components/common/DeferredSpinner';

const BillingForm = withCurrentUser(BillingFormShim);

interface Props {
  createOrganization: () => Promise<string>;
  deleteOrganization: () => void;
  onFreePlanChoose: () => Promise<void>;
  onPaidPlanChoose: () => void;
  onlyPaid?: boolean;
  open: boolean;
  startingPrice: string;
  subscriptionPlans: SubscriptionPlan[];
}

interface State {
  plan: Plan;
  ready: boolean;
  submitting: boolean;
}

export default class PlanStep extends React.PureComponent<Props, State> {
  mounted = false;

  constructor(props: Props) {
    super(props);
    this.state = {
      plan: props.onlyPaid ? Plan.Paid : Plan.Free,
      ready: false,
      submitting: false
    };
  }

  componentDidMount() {
    this.mounted = true;
    getExtensionStart('billing/billing').then(
      () => {
        if (this.mounted) {
          this.setState({ ready: true });
        }
      },
      () => {}
    );
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  handlePlanChange = (plan: Plan) => {
    this.setState({ plan });
  };

  stopSubmitting = () => {
    if (this.mounted) {
      this.setState({ submitting: false });
    }
  };

  handleFreePlanSubmit = () => {
    this.setState({ submitting: true });
    this.props.onFreePlanChoose().then(this.stopSubmitting, this.stopSubmitting);
  };

  renderForm = () => {
    const { submitting } = this.state;
    return (
      <div className="boxed-group-inner">
        {this.state.ready && (
          <>
            {!this.props.onlyPaid && (
              <PlanSelect
                onChange={this.handlePlanChange}
                plan={this.state.plan}
                startingPrice={this.props.startingPrice}
              />
            )}

            {this.state.plan === Plan.Paid ? (
              <BillingForm
                onCommit={this.props.onPaidPlanChoose}
                onFailToUpgrade={this.props.deleteOrganization}
                organizationKey={this.props.createOrganization}
                subscriptionPlans={this.props.subscriptionPlans}>
                {({ onSubmit, renderFormFields, renderSubmitGroup }) => (
                  <form onSubmit={onSubmit}>
                    {renderFormFields()}
                    <div className="billing-input-large big-spacer-top">
                      {renderSubmitGroup(
                        translate('onboarding.create_organization.create_and_upgrade')
                      )}
                    </div>
                  </form>
                )}
              </BillingForm>
            ) : (
              <div className="display-flex-center big-spacer-top">
                <SubmitButton disabled={submitting} onClick={this.handleFreePlanSubmit}>
                  {translate('my_account.create_organization')}
                </SubmitButton>
                {submitting && <DeferredSpinner className="spacer-left" />}
              </div>
            )}
          </>
        )}
      </div>
    );
  };

  render() {
    const stepTitle = translate(
      this.props.onlyPaid
        ? 'onboarding.create_organization.enter_payment_details'
        : 'onboarding.create_organization.choose_plan'
    );

    return (
      <Step
        finished={false}
        onOpen={() => {}}
        open={this.props.open}
        renderForm={this.renderForm}
        renderResult={() => null}
        stepNumber={2}
        stepTitle={stepTitle}
      />
    );
  }
}
