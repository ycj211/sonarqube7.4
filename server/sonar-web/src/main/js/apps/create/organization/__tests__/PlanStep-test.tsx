/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { shallow } from 'enzyme';
import PlanStep from '../PlanStep';
import { waitAndUpdate, click } from '../../../../helpers/testUtils';
import { Plan } from '../PlanSelect';

jest.mock('../../../../app/components/extensions/utils', () => ({
  getExtensionStart: jest.fn().mockResolvedValue(undefined)
}));

it('should render and use free plan', async () => {
  const onFreePlanChoose = jest.fn().mockResolvedValue(undefined);
  const wrapper = shallow(
    <PlanStep
      createOrganization={jest.fn().mockResolvedValue('org')}
      deleteOrganization={jest.fn().mockResolvedValue(undefined)}
      onFreePlanChoose={onFreePlanChoose}
      onPaidPlanChoose={jest.fn()}
      open={true}
      startingPrice="10"
      subscriptionPlans={[]}
    />
  );
  await waitAndUpdate(wrapper);
  expect(wrapper).toMatchSnapshot();
  expect(wrapper.dive()).toMatchSnapshot();

  click(wrapper.dive().find('SubmitButton'));
  expect(onFreePlanChoose).toBeCalled();
});

it('should upgrade', async () => {
  const onPaidPlanChoose = jest.fn();
  const wrapper = shallow(
    <PlanStep
      createOrganization={jest.fn().mockResolvedValue('org')}
      deleteOrganization={jest.fn().mockResolvedValue(undefined)}
      onFreePlanChoose={jest.fn().mockResolvedValue(undefined)}
      onPaidPlanChoose={onPaidPlanChoose}
      open={true}
      startingPrice="10"
      subscriptionPlans={[]}
    />
  );
  await waitAndUpdate(wrapper);

  wrapper
    .dive()
    .find('PlanSelect')
    .prop<Function>('onChange')(Plan.Paid);
  expect(wrapper.dive()).toMatchSnapshot();

  wrapper
    .dive()
    .find('Connect(withCurrentUser(BillingFormShim))')
    .prop<Function>('onCommit')();
  expect(onPaidPlanChoose).toBeCalled();
});

it('should preselect paid plan', async () => {
  const wrapper = shallow(
    <PlanStep
      createOrganization={jest.fn()}
      deleteOrganization={jest.fn().mockResolvedValue(undefined)}
      onFreePlanChoose={jest.fn().mockResolvedValue(undefined)}
      onPaidPlanChoose={jest.fn()}
      onlyPaid={true}
      open={true}
      startingPrice="10"
      subscriptionPlans={[]}
    />
  );
  await waitAndUpdate(wrapper);
  expect(wrapper).toMatchSnapshot();
  expect(wrapper.dive()).toMatchSnapshot();
});
