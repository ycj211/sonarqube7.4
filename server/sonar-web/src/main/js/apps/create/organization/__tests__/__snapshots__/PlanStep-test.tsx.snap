// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should preselect paid plan 1`] = `
<Step
  finished={false}
  onOpen={[Function]}
  open={true}
  renderForm={[Function]}
  renderResult={[Function]}
  stepNumber={2}
  stepTitle="onboarding.create_organization.enter_payment_details"
/>
`;

exports[`should preselect paid plan 2`] = `
<div
  className="boxed-group onboarding-step is-open"
>
  <div
    className="onboarding-step-number"
  >
    2
  </div>
  <div
    className="boxed-group-header"
  >
    <h2>
      onboarding.create_organization.enter_payment_details
    </h2>
  </div>
  <div
    className=""
  >
    <div
      className="boxed-group-inner"
    >
      <Connect(withCurrentUser(BillingFormShim))
        onCommit={[MockFunction]}
        onFailToUpgrade={[MockFunction]}
        organizationKey={[MockFunction]}
        subscriptionPlans={Array []}
      >
        <Component />
      </Connect(withCurrentUser(BillingFormShim))>
    </div>
  </div>
</div>
`;

exports[`should render and use free plan 1`] = `
<Step
  finished={false}
  onOpen={[Function]}
  open={true}
  renderForm={[Function]}
  renderResult={[Function]}
  stepNumber={2}
  stepTitle="onboarding.create_organization.choose_plan"
/>
`;

exports[`should render and use free plan 2`] = `
<div
  className="boxed-group onboarding-step is-open"
>
  <div
    className="onboarding-step-number"
  >
    2
  </div>
  <div
    className="boxed-group-header"
  >
    <h2>
      onboarding.create_organization.choose_plan
    </h2>
  </div>
  <div
    className=""
  >
    <div
      className="boxed-group-inner"
    >
      <PlanSelect
        onChange={[Function]}
        plan="free"
        startingPrice="10"
      />
      <div
        className="display-flex-center big-spacer-top"
      >
        <SubmitButton
          disabled={false}
          onClick={[Function]}
        >
          my_account.create_organization
        </SubmitButton>
      </div>
    </div>
  </div>
</div>
`;

exports[`should upgrade 1`] = `
<div
  className="boxed-group onboarding-step is-open"
>
  <div
    className="onboarding-step-number"
  >
    2
  </div>
  <div
    className="boxed-group-header"
  >
    <h2>
      onboarding.create_organization.choose_plan
    </h2>
  </div>
  <div
    className=""
  >
    <div
      className="boxed-group-inner"
    >
      <PlanSelect
        onChange={[Function]}
        plan="paid"
        startingPrice="10"
      />
      <Connect(withCurrentUser(BillingFormShim))
        onCommit={[MockFunction]}
        onFailToUpgrade={[MockFunction]}
        organizationKey={[MockFunction]}
        subscriptionPlans={Array []}
      >
        <Component />
      </Connect(withCurrentUser(BillingFormShim))>
    </div>
  </div>
</div>
`;
